### DISCUSSION TITLE
allow attributes & modifiers on a tuples fields

### DISCUSSION THREAD

**Discussion by lucasmeijer** [dotnet/csharplang#1149] [General]
We'd love to be able to write the following code:
    struct transform { public float x; }
    struct rigidbody { public float x; }

```C#
    static void Main()
    {
        EntityEnumerator<T> GetEntities<T>() =>/*.....*/

        foreach(var entity in GetEntities<(readonly ref transform t, rigidbody r)>()
        {
            entity.r.x++;
            entity.t.x++;  //<-- compiler error, field t of the tuple is a readonly field
        }
    }
```

In the same vain, we also have some attributes we'd love to be able to place:
```C#
           foreach(var entity in GetEntities<(transform t, [Prioritize] rigidbody r)>()
 ```
I realize we can make an ad-hoc struct, and put the readonly modifier or attribute on there, but it would be amazing to be able to use the concise new tuple syntax.

**Comment by svick** [comment]
What would that compile to? Tuples use the `System.ValueTuple` family of types, which means there is no way to place attributes on them.

And even with https://github.com/dotnet/csharplang/issues/1147, you would need a version of `ValueTuple` for all the possible combinations of `readonly` and `ref` if you wanted modifiers to work. I don't think that's realistic.

**Comment by iam3yal** [comment]
@lucasmeijer The [Records](https://github.com/dotnet/csharplang/issues/39) feature would give you more flexibility, it's not going to be tuples in anyway but you could at least put attributes on members but that's the future. :)
